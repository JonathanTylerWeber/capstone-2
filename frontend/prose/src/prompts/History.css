.history-h1 {
  font-size: 3em;
  font-weight: bold;
}

.history-container {
  margin-top: 10em;
}

.h-text {
  font-size: 1.2em;
  width: 100%;
  white-space: pre-wrap;
  word-wrap: break-word;
  box-sizing: border-box;
  margin-bottom: 2em;
}

.h-title {
  font-weight: bold;
}

.prompt {
  background-color: #214050;
  padding: 6px 12px;
  border-radius: 10px;
  margin-top: 2em;
  margin-bottom: 5em;
  padding: 2em 2em;
}

.trash {
  font-size: 1em;
  border-radius: 5px;
}