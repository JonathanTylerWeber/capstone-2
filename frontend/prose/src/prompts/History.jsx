import React, { useState, useEffect } from "react";
import ProseApi from '../API/api';
import "./History.css"
import { Container } from "react-bootstrap"
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import { faTrashCan } from "@fortawesome/free-solid-svg-icons";

function History({ currentUser }) {
  const [prompts, setPrompts] = useState([]);

  useEffect(() => {
    const fetchUserPrompts = async () => {
      try {
        let username = currentUser.username;
        let res = await ProseApi.getUserPrompts(username);
        setPrompts(res.prompts);
      } catch (error) {
        console.error("Failed to fetch user prompts:", error);
      }
    };

    fetchUserPrompts();
  }, [currentUser]);

  const handleDeletePrompt = async (promptId) => {
    try {
      await ProseApi.deletePrompt(promptId, currentUser.username);
      // Filter out the deleted prompt from the prompts array
      setPrompts(prompts.filter(prompt => prompt.id !== promptId));
    } catch (error) {
      console.error("Failed to delete prompt:", error);
    }
  };

  return (
    <Container className="history-container">
      <h1 className="history-h1">Prompt History:</h1>
      {prompts && prompts.length > 0 ? (
        prompts.map(prompt => (
          <div key={prompt.id} className="prompt">
            <h3 className="h-title">Type:</h3>
            <p className="h-text">{prompt.type}</p>
            <h3 className="h-title">Adjective:</h3>
            <p className="h-text">{prompt.adj}</p>
            <h3 className="h-title">Prompt:</h3>
            <pre className="h-text">{prompt.prompt}</pre>
            <h3 className="h-title">Rating:</h3>
            <p className="h-text">{prompt.rating}</p>
            <h3 className="h-title">Rewrite:</h3>
            <pre className="h-text">{prompt.rewrite}</pre>
            <button variant="primary" className="btn"
              onClick={() => handleDeletePrompt(prompt.id)}>
              <FontAwesomeIcon className="trash" icon={faTrashCan} />
            </button>
          </div>
        ))
      ) : (
        <p>No prompts available</p>
      )}
    </Container>
  );
}



{/* <p>{prompts.prompts[0].prompt}</p> */ }

export default History;