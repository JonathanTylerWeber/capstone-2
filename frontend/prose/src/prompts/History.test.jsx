import React from 'react';
import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import History from './History';
import ProseApi from '../API/api';

// Mock the API module
vi.mock('../API/api', async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...actual,
    getUserPrompts: vi.fn().mockResolvedValue({ prompts: [] }), // Mock the getUserPrompts function
    deletePrompt: vi.fn().mockResolvedValue({}), // Mock the deletePrompt function
  };
});

describe('History', () => {
  it('renders without crashing', async () => {
    render(<History />);
  });

  it("matches snapshot", function () {
    const { asFragment } = render(<History />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders no prompts available message when there are no prompts', async () => {
    // Mock the getUserPrompts function to return an empty array of prompts
    const getUserPromptsMock = vi.fn();
    getUserPromptsMock.mockResolvedValueOnce({ prompts: [] });
    ProseApi.getUserPrompts = getUserPromptsMock;

    render(<History currentUser={{ username: 'testuser' }} />);
    await waitFor(() => {
      expect(screen.getByText('No prompts available')).toBeInTheDocument();
    });
    expect(ProseApi.getUserPrompts).toHaveBeenCalledWith('testuser');
  });

  it('renders prompts and handles prompt deletion', async () => {
    // Mock prompts data
    // Mock the deletePrompt function to return an empty object
    const deletePromptMock = vi.fn();
    deletePromptMock.mockResolvedValueOnce({});
    ProseApi.deletePrompt = deletePromptMock;

    const prompts = [
      {
        id: 1,
        type: 'Test Type',
        adj: 'Test Adjective',
        prompt: 'Test Prompt',
        rating: 5,
        rewrite: 'Test Rewrite',
      },
    ];

    // Mock the getUserPrompts function to return the mock prompts
    ProseApi.getUserPrompts.mockResolvedValueOnce({ prompts });

    render(<History currentUser={{ username: 'testuser' }} />);

    // Wait for prompts to be rendered
    await waitFor(() => {
      expect(screen.getByText('Test Type')).toBeInTheDocument();
    });

    // Assert that the delete button exists and simulate click
    const deleteButton = screen.getByRole('button');
    expect(deleteButton).toBeInTheDocument();
    fireEvent.click(deleteButton);

    // Wait for deletion to be handled
    await waitFor(() => {
      expect(ProseApi.deletePrompt).toHaveBeenCalledWith(1, 'testuser');
    });

    // Ensure the prompt is no longer rendered after deletion
    expect(screen.queryByText('Test Type')).not.toBeInTheDocument();
  });
});
