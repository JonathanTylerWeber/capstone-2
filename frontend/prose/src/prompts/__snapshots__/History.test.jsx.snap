// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`History > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="history-container container"
  >
    <h1
      class="history-h1"
    >
      Prompt History:
    </h1>
    <p>
      No prompts available
    </p>
  </div>
</DocumentFragment>
`;
